Add KEP for default Even Pods Spreading (#1260)
kubernetes/enhancements: keps/sig-scheduling/20190926-default-even-pod-spreading.md

  - [Implementation Details](#implementation-details)
    - [In the metadata/predicates/priorities flow](#in-the-metadatapredicatespriorities-flow)
    - [In the scheduler framework](#in-the-scheduler-framework)
  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Alternatives [optional]](#alternatives-optional)
<!-- /toc -->

## Release Signoff Checklist

- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
- [ ] KEP approvers have set the KEP status to `implementable`
- [ ] Design details are appropriately documented
- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] Graduation criteria is in place
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Summary

> With `EvenPodsSpreading`, workload authors can define spreading rules for their loads based on the

Reviewer: Can we use runtime class to implement this feature, we could add another field in the `Scheduling` struct. https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/runtime-class-scheduling.md

Author: `RuntimeClass` the objective to deal with different runtimes in a node. This feature doesn't have to do anything with that.

- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Summary

With `EvenPodsSpreading`, workload authors can define spreading rules for their loads based on the
topology of the clusters. 

By introducing configurable default spreading rules, workloads can be spread in the topology of
the cluster according to opinionated rules set by a cluster operator, without having to explicitly
define them. And workloads with specific needs can still override the rules by defining them in
their PodSpec.

## Motivation

In order for a workload (pod) to use `EvenPodsSpreading`:

1. Authors have to have an idea of the underlying topology.
1. PodSpecs become less portable if their spreading rules are tailored to a specific topology.

On the other hand, cluster operators know the underlying topology of the cluster, which makes
them suitable to provide default spreading rules for all workloads in their cluster.

### Goals

> - Cluster operators can define default spreading rules for pods that don't provide any.

Reviewer 2: ```suggestion
- Cluster operators can define default spreading rules for pods that don't provide any `pod.spec.topologySpreadConstraints`.
```

> - Workloads are spread with the default rules if they belong to the same service, controller,

Reviewer 2: "controller" is vague, I checked the node and it here means "replication controller".

Reviewer 3: yeah, controller refers to replication controller, which is the predecessor of ReplicaSet.

replica set or stateful set,
and if they don't define `TopologySpreadConstraints`.

### Non-Goals

- Removal of `SelectorSpreadPriority`, `ServiceSpreadingPriority` or `ServiceAntiAffinity` priorities.

## Proposal

### User Stories

#### Story 1

As a cluster operator, I want to set default spreading rules for workloads in the cluster.

> Currently, `SelectorSpreadPriority` provides a canned priority that spreads across nodes

Reviewer 2: "nodes and zones", or just "zones"?

and zones (`failure-domain.beta.kubernetes.io/zone`). However, the nodes in my cluster have
custom topology keys (for physical host, rack, etc.).

#### Story 2

As a workload author, I want to spread the workload in the cluster, but:
(1) I don't know the topology of the cluster I'm running on.
(2) I want to be able to run my PodSpec in different clusters (on-prem and cloud).

### Implementation Details/Notes/Constraints

> Note that a priority given by default `EvenPodsSpreading` rules could conflict with

Reviewer 2: Are you trying to say "a topologySpreadConstraint with field WhenUnsatisfiable = ScheduleAnyway"?

`SelectorSpreadingPriority`.

> Operators can disable `SelectorSpreadingPriority`. But once default rules for `EvenPodsSpreading` is GA,

Reviewer 2: I don't follow what default rules for `EvenPodsSpreading` is GA? There are no "default" rules for `EvenPodsSpreading` actually.

Reviewer 2: I think when you refer to rules/constraints, we should use the term "topology spread constraints". And only use `EvenPodsSpreading` when referring to the alpha feature we implemented in 1.15.

we can consider removing `SelectorSpreadingPriority` and replacing it by an equivalent

> k8s default to the default rules for `EvenPodsSpreading`.

Reviewer 3: I think we want to present this KEP as the replacement for the SelectorSpreadingPriority, they shouldn't exist together. To make the transition frictionless, we can set a default value for what we are proposing here such that the scheduler's default behavior isn't changed (i.e., we set a default configuration for "default even pods spreading" to spread across zones).

Reviewer 3: What I meant is that we shouldn't allow configuring both at the same time at all. The paragraph is still talking about the possibility of that being the case.

Currently, SelectorSpreadingPriority is used as one of the priority functions when no Policy is specified, and it is a value that we accept in the Policy file. Those two aspects are our guarantees to the users. We can solve them by doing the following if this feature is enabled:

1) SelectorSpreadingPriority is removed from the default set of priorities. The default configuration for cluster-wide EvenPodsSpreading is set such that we maintain the same current scheduler default behavior that SelectorSpreadingPriority offers (or as close as possible).

2) During the deprecation period of Policy (and by association SelectorSpreadingPriority), we say when set, SelectorSpreadingPriority overrides cluster-wide EvenPodsSpreading.

WDYT?

Author: I agree that they are conflicting. See updated section.

### Risks and Mitigations

`EvenPodsSpreading` has some overhead and we currently ensure that pods that don't use the

> feature get minimally affected. After default rules for `EvenPodsSpreading` is rolled out,

Reviewer 2: Ditto. What does "default rules" mean? Are you trying to mean **this feature** with partial APIs of `EvenPodsSpreading`? If it's the case, we should explicitly reword to the API "topologySpreadConstraint".

all pods will run through the algorithms.
However, we should ensure that the running overhead is not significantly higher than
`SelectorSpreadingPriority` if using the k8s default.

## Design Details

### API

> A new structure called `TopologySpreadConstraint` is introduced to `KubeSchedulerConfiguration`.

Reviewer 2: ```suggestion
`TopologySpreadConstraint` or `DefaultTopologySpreadConstraints`?
```

```go
type KubeSchedulerConfiguration struct {
	....
	// DefaultTopologySpreadConstraints defines spreading constraints to be applied to pods
	// that don't define any.

> 	// If not specified, the scheduler applies the following default.

Reviewer 3: What are you referring to with "If not specified"? I think the whole sentence in this line is redundant.

	// +optional

> 	DefaultTopologySpreadConstraints []TopologySpreadConstraint

Reviewer 3: I would drop the prefix Default, I think it is not necessary. It will be confusing when we start talking about a default value for DefaultTopologySpreadConstraints.

Reviewer 3: Component config is a cluster-wide configuration by definition, adding a "Default" prefix does not really clarify the semantics of this configuration variable as you state it, at the end of the day it has to be explained in the docs/comments that this value can be overridden at the pod level.

Author: See updated name and comment.

	....
}

// TopologySpreadConstraint specifies how to spread pods among the given topology.
// Pod selectors are deduced from the resource definitions that the pod belongs to
// (includes services, controllers, replica sets and stateful sets).
type TopologySpreadConstraint struct {
	MaxSkew int32
	TopologyKey string
	WhenUnsatisfiable corev1.UnsatisfiableConstraintAction
}
```

> Note that `TopologySpreadConstraint` is similar to `k8s.io/api/core/v1.TopologySpreadConstraint`,

Reviewer 2: As current `TopologySpreadConstraint.LabelSelector` is optional, you can reuse that structure. (validation logic can help detect illegal presence of labelSelector)

If you still want to create a new structure, the name should be different from `TopologySpreadConstraint` to avoid confusion.

Reviewer 3: > As current `TopologySpreadConstraint.LabelSelector` is optional, you can reuse that structure. (validation logic can help detect illegal presence of labelSelector)

@Huang-Wei  What happens currently in EvenPodsSpreading when the label selector is not specified?

Author: These are different APIs, so I think they can be named the same. But I'm happy to reuse the struct if the selector is indeed optional. And during validation we ensure that it was not set.

Reviewer 2: @ahg-g If `LableSelector` is nil, no pods will be matched, then the incoming pod can land on any node which has the topologyKey(s) present. So in other words, it doesn't fit this KEP's requirement. But we _can_ modify its semantics and document it.

except that it doesn't define selectors.

### Default rules

These will be the default rules for the cluster when the operator doesn't provide any:

> ```yaml

Reviewer 2: ```yaml
defaultTopologySpreadConstraints:
- maxSkew: 1
  topologyKey: "kubernetes.io/hostname"
  whenUnsatisfiable: ScheduleAnyway
- maxSkew: 1
  topologyKey: "failure-domain.beta.kubernetes.io/zone"
  whenUnsatisfiable: ScheduleAnyway
```

Author: Done

> defaultTopologySpreadConstraints:

Reviewer 2: Ditto. Please prettify the yaml format.

Author: Done

  -
    maxSkew: 1
    topologyKey: "kubernetes.io/hostname"
    whenUnsatisfiable: ScheduleAnyway
  -
    maxSkew: 1
    topologyKey: "failure-domain.beta.kubernetes.io/zone"
    whenUnsatisfiable: ScheduleAnyway
```

### How user stories are addressed

Let's say we have a cluster that has a topology based on physical hosts and racks.
Then, an operator can set the following scheduler configuration:

```yaml
apiVersion: componentconfig/v1alpha1
defaultTopologySpreadConstraints:
  -
    maxSkew: 5
    topologyKey: "example.com/topology/physical_host"
    whenUnsatisfiable: ScheduleAnyway
  -
    maxSkew: 15
    topologyKey: "example.com/topology/rack"
    whenUnsatisfiable: DoNotSchedule
```

Then, a workload author could have the following `ReplicaSet`:

```yaml
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: replicated_demo
spec:
  replicas: 3
  selector:
    matchLabels:
      app: demo
  template:
    metadata:
      labels:
        app: demo
    spec:
      containers:
      - name: php-redis
        image: example.com/registry/demo:latest
```

> Note that the workload author didn't provide a spreading rules.

Reviewer 2: ```suggestion
Note that the workload author didn't provide spreading rules in the `pod.spec`.
```

Author: Done

> The following rules will be applied to the pods of this replica set before running the

Reviewer 2: Suggest rewording to "The following spreading rules defined in ComponentConfig will be derived and applied in runtime."

Author: Done

algorithms for Even Pods Spreading:

> ```yaml

Reviewer 2: Please prettify the yaml.

Author: Done

topologySpreadConstraints:
  -
    maxSkew: 5
    TopologyKey: "example.com/topology/physical_host"
    WhenUnsatisfiable: ScheduleAnyway
    selector:
      matchLabels:
        app: demo
  -
    maxSkew: 15
    TopologyKey: "example.com/topology/rack"
    WhenUnsatisfiable: DoNotSchedule
    selector:
      matchLabels:
        app: demo
```

> These rules are internal to the scheduler and they don't get reflected in the apiserver.

Reviewer 2: Suggest rewording to "Please note that these rules are honored internally in the scheduler, and they are NOT persisted into PodSpec via API Server."

Author: Done

### Implementation Details

#### In the metadata/predicates/priorities flow

1. Calculate the spreading constraints for the pod as part of the metadata calculation.
   Use the constraints provided by the pod or calculate the default ones if they don't provide any.
1. When running the predicates or priorities, use the constraints stored in the metadata.

#### In the scheduler framework

1. Calculate spreading constraints for the pod in the `PreFilter` extension point. Store them
   in the `PluginContext`.
1. In the `Filter` and `Score` extension points, use the stored spreading constraints instead of
   the ones defined by the pod.

### Test Plan

To ensure this feature to be rolled out in high quality. Following tests are mandatory:

- **Unit Tests**: All core changes must be covered by unit tests.
- **Integration Tests**: One integration test for the default rules and one for custom rules.
- **Benchmark Tests**: A benchmark test that compare the default rules against `SelectorSpreadingPriority`.
  The performance should be as close as possible.

### Graduation Criteria

Alpha (v1.17):

> [ ] Scheduler Component Config API changes.

Reviewer 2: ```suggestion
- [ ] Scheduler Component Config API changes.
```

And also apply to the items below.

Author: Done

### Implementation Details/Notes/Constraints

#### Relationship with SelectorSpreadingPriority

Note that Default `topologySpreadConstraints` has a similar effect to `SelectorSpreadingPriority`.
Given that the latter is not configurable, they could return conflicting priorities, which
may not be the intention of an operator. On the other hand, a proper Default for `topologySpreadConstraints`
could provide the same priority as `SelectorSpreadingPriority`. Thus, there's no need for the
features to co-exist.

Give that we guard Default `topologySpreadConstraints` behind a feature flag,
these would be its semantics:

- If the feature is enabled, `SelectorSpreadingPriority` is removed from the default set of priorities.
  K8s provides the Default `topologySpreadConstraints` that matches the priority given by
  `SelectorSpreading` if the cluster operator doesn't specify one.
- If the cluster operator provides a Policy that includes `SelectorSpreadingPriority` and
  `EvenPodsSpreadPriority`, K8s provides an empty Default `topologySpreadConstraints`.
  The cluster operator can still specify Default `topologySpreadConstraints`,

>   in which case both priorities run.

Reviewer 3: I don't we should run both priorities, I think we should clarify that if both are specified, then one overrides the other.

Author: So your take would be that `EvenPodsSpreadPriority` overrides and disables `SelectorSpreadingPriority`?

Reviewer 3: We should have a single implementation, a plugin. This plugin will either be configured using the priority-to-plugin translation logic we have if ```SelectorSpreadingPriority``` is specified, otherwise it is configured using the new ComponentConfig parameters you will be adding.

Author: See updated clarifications

### Implementation Details/Notes/Constraints

#### Relationship with SelectorSpreadingPriority

Note that Default `topologySpreadConstraints` has a similar effect to `SelectorSpreadingPriority`.
Given that the latter is not configurable, they could return conflicting priorities, which
may not be the intention of an operator. On the other hand, a proper Default for `topologySpreadConstraints`
could provide the same priority as `SelectorSpreadingPriority`. Thus, there's no need for the
features to co-exist.

K8s will set Default `topologySpreadConstraints` and remove `SelectorSpreadingPriority`
from the k8s providers (`DefaultProvider` and `ClusterAutoscalerProvider`). The set
[default](#default-rules) will have a similar effect.

If an operator sets a Policy, these are the semantics of the presence of `SelectorSpreadingPriority`
and/or `EvenPodsSpreadPriority`:

| SelectorSpreading | EvenPodsSpread | Valid | Pod spread constraints                    |
| :---------------: | :------------: | :---: | :---------------------------------------: |
| N                 | Y              | Yes   | provided or [k8s default](#default-rules) |

> | Y                 | Y              | Yes   | provided or [k8s default](#default-rules) |

Reviewer 3: the constraint should be *k8s default* always, never in this case because SelectorSpreading takes precedence.

Reviewer 3: or you can consider specifying both an invalid configuration.

Author: I went with the second option

      matchLabels:
        app: demo
```

Please note that these constraints are honored internally in the scheduler, but they are NOT
persisted in the PodSpec via API Server.

### Implementation Details

#### In the metadata/predicates/priorities flow

1. Calculate the spreading constraints for the pod as part of the metadata calculation.
   Use the constraints provided by the pod or calculate the default ones if they don't provide any.
1. When running the predicates or priorities, use the constraints stored in the metadata.

#### In the scheduler framework

1. Calculate spreading constraints for the pod in the `PreFilter` extension point. Store them
   in the `PluginContext`.
1. In the `Filter` and `Score` extension points, use the stored spreading constraints instead of
   the ones defined by the pod.

### Test Plan

To ensure this feature to be rolled out in high quality. Following tests are mandatory:

- **Unit Tests**: All core changes must be covered by unit tests.
- **Integration Tests**: One integration test for the default rules and one for custom rules.
- **Benchmark Tests**: A benchmark test that compare the default rules against `SelectorSpreadingPriority`.

>   The performance should be as close as possible.

Reviewer 4: Sorry for jumping late into this discussion - but do we believe we can do that efficiently?
I'm asking in the context of large clusters - i.e. 5k nodes.

I want to ensure that we won't introduce a visible regression, because this may become a blocker.

Reviewer 3: I think we should be able to do this as efficiently as the current logic, I don't see why we can't. I agree that it is a blocker if this will be causing performance regressions though.

Author: `SelectorSpreadingPriority` is currently part of the default algorithms that run. Theoretically, we are talking about the same problem, and then it should be possible to devise an algorithm for `EvenPodsSpreadPriority` with the same complexity.
We will be backing it with performance measurements, as seen in the Test Plan.

Reviewer 3: @alculquicondor  perhaps we should explicitly call that out in the KEP?

Author: The Test Plan is part of the graduation criteria.

### Graduation Criteria

> Alpha (v1.17):

Reviewer 3: we should say that this feature is tired to even pods spread, so they will graduate together.
